cli/src/main/java/org/jboss/as/cli/handlers/ModuleNameTabCompleter.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2016, JBoss Inc., and individual contributors as indicated
+ * by the @authors tag.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jboss.as.cli.handlers;
+
+import org.jboss.as.cli.CommandContext;
+import org.jboss.as.cli.CommandLineCompleter;
+import org.jboss.as.cli.EscapeSelector;
+import org.jboss.as.cli.Util;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * @author Bartosz Spyrko-Smietanko
+ */
+public class ModuleNameTabCompleter implements CommandLineCompleter {
+
+    private static final EscapeSelector ESCAPE_SELECTOR = ch -> ch == '\\' || ch == ' ' || ch == '"';
+    private static final char MODULE_NAME_SEPARATOR = '.';
+    public static final String LAYERS_DIR = "system/layers";
+
+    private final File modulesRoot;
+    private final File layersDir;
+
+    public ModuleNameTabCompleter(String modulesDir) {
+        modulesRoot = new File(modulesDir);
+        layersDir = new File(modulesDir, LAYERS_DIR);
+    }
+
+    @Override
+    public int complete(CommandContext ctx, String buffer, int cursor, List<String> candidates) {
+        final String userEntry = buffer == null ? "" : buffer;
+        final Set<String> suggestions = new TreeSet<>(); // TreeSet deals with duplication and ordering
+
+        List<File> moduleTrees = findInitialModelDirectories();
+
+        moduleTrees.forEach(f -> findSuggestion(f, f.getName(), userEntry, suggestions));
+
+        candidates.addAll(suggestions);
+
+        return 0;
+    }
+
+    private List<File> findInitialModelDirectories() {
+        List<File> moduleTrees = new ArrayList<>();
+
+        moduleTrees.addAll(Arrays.asList(modulesRoot.listFiles(f -> isNotLayersFolder(f))));
+
+        if (layersDir.exists()) {
+            for (File layer : layersDir.listFiles(File::isDirectory)) {
+                moduleTrees.addAll(Arrays.asList(layer.listFiles(f -> isNotLayersFolder(f))));
+            }
+        }
```

Comment:
This needs updating.

The valid locations that may be the first level of a module directory tree are:

1) Child dirs of the module root that are not named 'system'. That's line 67 above.
2) Child dirs of any dir whose parent dir is <modulesDir>/system/layers. So the 'isNotLayersFolder' check on L71 is not correct as <modulesDir>/system/layers/system would be a valid location if someone produced a layer named 'system'.
3) Child dirs of any dir whose parent dir is <modulesDir>/system/addons.